feat: Booking EmailAndSms Notifications Tasker

ThyMinimalDev opened the pull request:
## What does this PR do?

Adds async email and SMS sending for booking events using Trigger.dev with a safe sync fallback to reduce request time and improve reliability. Integrates the tasker into RegularBookingService via DI, aligns logging with tslog, and loads Trigger.dev tasks dynamically to avoid test failures.

    New Features
        Added Tasker abstraction with a primary Trigger.dev tasker (toggled by ENABLE_ASYNC_TASKER) and a sync fallback.
        Added Trigger.dev tasks and config within packages/features: request, confirm, reschedule, rr-reschedule, plus a simple logger.
        Hooked booking flow to dispatch tasks via BookingEmailAndSmsTasker; notifications are sent via BookingEmailSmsHandler.
        Introduced base Tasker utilities and DI modules to wire the tasker/handler into API services.
        Updated logger bridge to match tslog’s interface.
        Switched Trigger.dev task imports to dynamic to avoid loading the SDK during tests.
        Added team feature flag "booking-email-sms-tasker" to control rollout.
        Extended booking fetch to include attendee phone numbers and seat metadata for SMS.
        Added booking-notifications queue and retry policy; included trigger.config.ts and dev/deploy scripts.

    Migration
        Set TRIGGER_SECRET_KEY and TRIGGER_API_URL; optionally set ENABLE_ASYNC_TASKER=true; run Trigger.dev via packages/features (yarn dev:trigger / yarn deploy:trigger).
        Optionally set LOGGER_BRIDGE_LOG_LEVEL to control API logging.
        Optionally enable the "booking-email-sms-tasker" feature flag.
        No changes needed if Trigger.dev isn’t configured—the sync fallback will handle notifications.

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.





































































































































<!-- This is an auto-generated description by cubic. -->
---
## Summary by cubic
Adds async email and SMS sending for booking events using Trigger.dev with a safe sync fallback to reduce request time and improve reliability. Integrates the tasker into RegularBookingService via DI, aligns logging with tslog, and loads Trigger.dev tasks dynamically to avoid test failures.

- **New Features**
  - Added Tasker abstraction with a primary Trigger.dev tasker (toggled by ENABLE_ASYNC_TASKER) and a sync fallback.
  - Added Trigger.dev tasks and config within packages/features: request, confirm, reschedule, rr-reschedule, plus a simple logger.
  - Hooked booking flow to dispatch tasks via BookingEmailAndSmsTasker; notifications are sent via BookingEmailSmsHandler.
  - Introduced base Tasker utilities and DI modules to wire the tasker/handler into API services.
  - Updated logger bridge to match tslog’s interface.
  - Switched Trigger.dev task imports to dynamic and moved heavy handler/workflow imports to runtime to reduce cold start and avoid test/circular import issues.
  - Added team feature flag "booking-email-sms-tasker" to control rollout.
  - Extended booking fetch to include attendee phone numbers and seat metadata for SMS.
  - Added booking-notifications queue and retry policy; included trigger.config.ts and dev/deploy scripts.

- **Migration**
  - Set TRIGGER_SECRET_KEY and TRIGGER_API_URL; optionally set ENABLE_ASYNC_TASKER=true; run Trigger.dev via packages/features (yarn dev:trigger / yarn deploy:trigger).
  - Set TRIGGER_DEV_PROJECT_REF for Trigger.dev (used by packages/features/trigger.config.ts).
  - Optionally set LOGGER_BRIDGE_LOG_LEVEL to control API logging.
  - Optionally enable the "booking-email-sms-tasker" feature flag.
  - No changes needed if Trigger.dev isn’t configured—the sync fallback will handle notifications.

<sup>Written for commit 974fb49971012d3bf7669285c577c44b17068248. Summary will update automatically on new commits.</sup>

<!-- End of auto-generated description by cubic. -->

ThyMinimalDev committed:
wip

ThyMinimalDev committed:
wip

ThyMinimalDev committed:
feature: Booking Tasker without DI yet

ThyMinimalDev commented on packages/features/bookings/lib/tasker/BookingTasker.ts:
```diff
@@ -0,0 +1,47 @@
+  logger: ILogger;
+}
+
+export class BookingTasker extends Tasker<IBookingTasker> {
```
TODO: rename all taskers to reflect they are specialized for SMS and Emails

hbjORbj reviewed (commented):
Code looks great, tests and typechecks need to be fixed though. Will come back later for a 2nd review and some testing.

ThyMinimalDev committed:
feature: Booking Tasker with DI

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

hbjORbj committed:
fix type check 1

hbjORbj committed:
fix type check 2

hbjORbj committed:
fix

ThyMinimalDev committed:
comment booking tasker for now

ThyMinimalDev committed:
fix: DI regularBookingService api v2

devin-ai-integration[bot] committed:
fix: convert trigger.dev SDK imports to dynamic imports to fix unit tests

The unit tests were failing because BookingEmailAndSmsTriggerTasker.ts had static imports of trigger files that depend on @trigger.dev/sdk. This caused Vitest to try to resolve the SDK at module load time, even though it should be optional.

Changed all imports in BookingEmailAndSmsTriggerTasker.ts from static to dynamic (using await import()) so the trigger files are only loaded when the tasker methods are actually called, not at module load time during tests.

This fixes the 'Failed to load url @trigger.dev/sdk' errors that were causing 28+ test failures.

Co-Authored-By: morgan@cal.com <morgan@cal.com>

hbjORbj committed:
fix unit tests

ThyMinimalDev committed:
keep inline smsAndEmailHandler.send calls

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
chore: add team feature flag

ThyMinimalDev committed:
add satisfies ModuleLoader

ThyMinimalDev committed:
fix type check app flags

ThyMinimalDev committed:
move trigger in feature

hbjORbj committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
fix: add trigger.dev prisma  generator

ThyMinimalDev committed:
fix: email app statuses

ThyMinimalDev committed:
fix: CalEvtBuilder unit test

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
chore: improvements, schema, config, retry

ThyMinimalDev committed:
fixup! chore: improvements, schema, config, retry

ThyMinimalDev committed:
chore: cleanup code

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
chore: cleanup code

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
chore: clean code and give full payload

ThyMinimalDev committed:
remove log

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
add booking notifications queue

ThyMinimalDev committed:
add attendee phone number for sms

ThyMinimalDev committed:
bump trigger to 4.1.0

ThyMinimalDev committed:
add missing booking seat data in attendee

ThyMinimalDev committed:
update config

ThyMinimalDev committed:
fix logger regular booking service

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
fix: prisma as external deps of trigger

ThyMinimalDev committed:
fix yarn.lock

ThyMinimalDev committed:
revert change to example app booking page

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

devin-ai-integration[bot] committed:
fix: resolve circular dependencies and improve cold start performance in trigger tasks

- Convert BookingRepository import to type-only in CalendarEventBuilder.ts to eliminate circular dependency risk
- Convert EventNameObjectType, CalendarEvent, and JsonObject imports to type-only in BookingEmailAndSmsTaskService.ts
- Use dynamic imports in all trigger notification tasks (confirm, request, reschedule, rr-reschedule) to reduce cold start time
- Move heavy imports (BookingEmailSmsHandler, BookingRepository, prisma, TriggerDevLogger, BookingEmailAndSmsTaskService) inside run functions
- Eliminates module-level prisma import which violates repo guidelines and adds cold start overhead
- Reduces initial module dependency graph by deferring heavy imports (email templates, workflows, large repositories) until task execution

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

devin-ai-integration[bot] committed:
fix: improve cold start performance in reminderScheduler with dynamic imports

- Remove module-level prisma import (violates 'No prisma outside repositories' guideline)
- Use dynamic imports for UserRepository (1,168 lines) - only loaded when needed in EMAIL_ATTENDEE action
- Use dynamic imports for twilio provider (386 lines) - only loaded in cancelScheduledMessagesAndScheduleEmails
- Use dynamic imports for all manager functions by action type:
  - scheduleSMSReminder (387 lines) - loaded only for SMS actions
  - scheduleEmailReminder (459 lines) - loaded only for Email actions
  - scheduleWhatsappReminder (266 lines) - loaded only for WhatsApp actions
  - scheduleAIPhoneCall (478 lines) - loaded only for AI phone call actions
- Use dynamic imports for sendOrScheduleWorkflowEmails in cancelScheduledMessagesAndScheduleEmails
- Significantly reduces cold start time by deferring heavy module loading until execution paths need them
- Eliminates module-level prisma import that violated repository pattern guidelines

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
fix: improve cold start performance in BookingEmailSmsHandler with dynamic imports

- Remove module-level imports of all email-manager functions (653 LOC + 30+ email templates)
- Add dynamic imports in each method (_handleRescheduled, _handleRoundRobinRescheduled, _handleConfirmed, _handleRequested, handleAddGuests)
- Defer heavy email-manager loading until method execution
- Verified no circular dependencies between email-manager and bookings
- Significantly reduces cold start time for RegularBookingService and BookingEmailAndSmsTaskService

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
fix: use dynamic imports

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
update yarn lock

ThyMinimalDev committed:
code review

ThyMinimalDev committed:
trigger config project ref in env

hbjORbj reviewed (dismissed):
@ThyMinimalDev  will do some testing in QA and then merge this

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
update yarn lock

ThyMinimalDev committed:
add .env.example trigger variables

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
add .env.example trigger variables

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
fix: cleanup error handling and loggin

ThyMinimalDev committed:
fix: trigger config from env

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
fix: small typo fix

ThyMinimalDev committed:
fix: ai review comments

ThyMinimalDev committed:
fix: ai review comments

ThyMinimalDev committed:
ai review

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
prettier

ThyMinimalDev committed:
fix: atoms build

emrysal reviewed (approved):
LGTM! 🚀 🥳
